### DISCUSSION TITLE
scope when store is actionless

### DISCUSSION THREAD

**Discussion by K3TZR** [pointfreeco/swift-composable-architecture#1008] [Q&A]
I have a view that displays a list of Meters. It is "display only", i.e. no actions occur in this view. Here is the code of the main view followed by the Meter row view.

struct MeterView: View {
  let store: Store<ApiState, ApiAction>
  let sliceId: ObjectId?
  
  var body: some View {
    
    WithViewStore(store) { viewStore in
      VStack(alignment: .leading) {
        ForEachStore(
          store.scope(
            state: \.objects.meters,
            action: ApiAction.meter(id:action:)
          ),
          content: { MeterRowView(store: $0.actionless, sliceId: sliceId) }
        )
      }
      .foregroundColor(.secondary)
      .onAppear() { viewStore.send(.startMetersSubscription) }
      .onDisappear() { viewStore.send(.stopMetersSubscription) }
    }
  }
}

struct MeterRowView: View {
  let store: Store<Meter, Never>
  let sliceId: SliceId?
......

My question is how to express the scope statement without supplying an action. The action will never occur and having to supply it here (to please the compiler) requires me to create a useless action in ApiAction  ( case meter(id: MeterId, action: MeterAction) ) and to define a useless value for MeterState ( public enum MeterAction { case anything } ).

I've tried a number of variations but can't seem to get it right. I wasn't able to find any examples other than a mention in the discussions that the Action parameter of the scope method is optional.

**Comment by mbrandonw** [comment]
Hi @K3TZR, there's a property on `Store` called `actionless` that does just that:

https://github.com/pointfreeco/swift-composable-architecture/blob/12ae834de95e9b78ad74634456c70e9b199e8f3e/Sources/ComposableArchitecture/Store.swift#L402-L406

**Reply by K3TZR** [reply]
Hi Brandon, thank you for the speedy reply.

I'm aware of that and I've used it in the content line:

    ForEachStore(
      store.scope(
        state: \.objects.meters,
        action: ApiAction.meter(id:action:)
      ),
      **content: { MeterRowView(store: $0.actionless, sliceId: sliceId) }**
    )

but in the scope statement I can't seem to express the same sort of idea. If I remove the:

        **action: ApiAction.meter(id:action:)**

I end up with a number of compile errors. It's this line that requires the useless MeterAction.

**Comment by mbrandonw** [comment]
Ah I see. If each row of the view doesn't need actions, perhaps you don't even need a `ForEachStore`. You could just get away with a `ForEach` and passing plain data down to `MeterRowView` rather than a store. Would that work?

**Reply by K3TZR** [reply]
I just tried that, here's the modified code:

struct MeterView: View {
  let store: Store<ApiState, ApiAction>
  let sliceId: ObjectId?
  
  var body: some View {
    
    WithViewStore(store) { viewStore in
      VStack(alignment: .leading) {
        ForEach(viewStore.objects.meters ) { meter in
          MeterRowView(meter: meter, sliceId: sliceId)
        }
        
//        ForEachStore(
//          store.scope(
//            state: \.objects.meters,
//            action: ApiAction.meter(id:action:)
//          ),
//          content: { MeterRowView(store: $0.actionless, sliceId: sliceId) }
//        )
      }
      .foregroundColor(.secondary)
      .onAppear() { viewStore.send(.startMetersSubscription) }
      .onDisappear() { viewStore.send(.stopMetersSubscription) }
    }
  }
}

struct MeterRowView: View {
  let meter: Meter
  //  let store: Store<Meter, Never>
  let sliceId: SliceId?

It seems to have introduced a different problem. The MeterRow View is shown but it doesn't update any longer. The update was being triggered by the .onAppear which starts a subscription (effect) that listens to a publisher of changes to meters. For some reason, that doesn't work in this version. I'll have to debug a bit.

My guess is that my code is full of places where I've done things in a less than optimum way. This is my first venture into TCA. I'm rewriting an older app using TCA as a learning experience. The learning curve is pretty steep.

**Reply by K3TZR** [reply]
To be clearer, the "meters" are in an IdentifiedArrayOf<Meter>. I added a property to ApiState called "forceUpdate". It is toggled when I received an action that results from the meter value changing. It's crude in that it forces a redraw of all of the view not just the Meter portion but I couldn't see any other way to do it.

Apparently, using ForEach make the drawing of the MeterRowView a more static operation, it's only being done once and doesn't redraw when the forceUpdate property is toggled.

**Reply by mbrandonw** [reply]
I would need to see more code to know what is going on, like the `ApiState` and reducer.

Any changes to state held in the view store will cause a view to re-render, so if `forceUpdate` is being observed it should do what you want.

**Reply by K3TZR** [reply]
I found a way to make it work. I took the contents of the MeterRowView and placed it inside the ForEach in the MeterView, i.e. I no longer have a separate view for MeterRowView. That works, the meter values are updated (actually the entire view and all of the views upstream of it are updated) each time forceUpdate is toggled.

I'm not sure exactly why this works but maybe I'll understand it better after more experience with TCA / SwiftUI.

Thank you for your time and help.

Doug
